pkg/chunk/chunk_store.go

```diff
@@ -205,15 +205,16 @@ func (c *store) LabelValuesForMetricName(ctx context.Context, userID string, fro
 	}
 
 	var result []string
+	values := map[string]struct{}{}
 	for _, entry := range entries {
 		_, labelValue, _, _, err := parseChunkTimeRangeValue(entry.RangeValue, entry.Value)
 		if err != nil {
 			return nil, err
 		}
-		result = append(result, string(labelValue))
+		result = appendUniqueStrings(result, values, string(labelValue))
```

Comment:
Since this change is (now) unrelated to the bugfix, I would prefer it goes in a different commit with its own explanation.

You may want to make the same change in `parseIndexEntries()`, which would let you remove `uniqueStrings()`.

It's worth checking the benchmark numbers and/or making sure the benchmarks are using a realistic data set.

Reply:
good catch ! I'll remove it.

Reply:
voila !

Reply:
before :

```
pkg: github.com/cortexproject/cortex/pkg/chunk
BenchmarkParseIndexEntries500-16      	    4562	    257222 ns/op	   96427 B/op	    1503 allocs/op
BenchmarkParseIndexEntries2500-16     	     847	   1366432 ns/op	  482388 B/op	    7503 allocs/op
BenchmarkParseIndexEntries10000-16    	     198	   5846147 ns/op	 1928891 B/op	   30005 allocs/op
BenchmarkParseIndexEntries50000-16    	      38	  30553792 ns/op	 9607698 B/op	  150006 allocs/op
PASS
ok  	github.com/cortexproject/cortex/pkg/chunk	7.953s
```

after:

```
pkg: github.com/cortexproject/cortex/pkg/chunk
BenchmarkParseIndexEntries500-16      	    3678	    318019 ns/op	  130810 B/op	    1522 allocs/op
BenchmarkParseIndexEntries2500-16     	     747	   1632040 ns/op	  612466 B/op	    7563 allocs/op
BenchmarkParseIndexEntries10000-16    	     172	   6779048 ns/op	 2441477 B/op	   30219 allocs/op
BenchmarkParseIndexEntries50000-16    	      31	  36258715 ns/op	11602340 B/op	  151867 allocs/op
PASS
ok  	github.com/cortexproject/cortex/pkg/chunk	8.108s
```

Reply:
Adding half duplicate in the mix:

before:

```
pkg: github.com/cortexproject/cortex/pkg/chunk
BenchmarkParseIndexEntries500-16      	    4485	    262828 ns/op	   96346 B/op	    1503 allocs/op
BenchmarkParseIndexEntries2500-16     	     838	   1379892 ns/op	  482316 B/op	    7503 allocs/op
BenchmarkParseIndexEntries10000-16    	     201	   5907672 ns/op	 1928821 B/op	   30005 allocs/op
BenchmarkParseIndexEntries50000-16    	      39	  30373427 ns/op	 9607658 B/op	  150007 allocs/op
PASS
ok  	github.com/cortexproject/cortex/pkg/chunk	8.056s
```

after:
```
pkg: github.com/cortexproject/cortex/pkg/chunk
BenchmarkParseIndexEntries500-16      	    4454	    265512 ns/op	  109322 B/op	    1514 allocs/op
BenchmarkParseIndexEntries2500-16     	     878	   1356005 ns/op	  526739 B/op	    7536 allocs/op
BenchmarkParseIndexEntries10000-16    	     211	   5586487 ns/op	 2107155 B/op	   30116 allocs/op
BenchmarkParseIndexEntries50000-16    	      37	  29155700 ns/op	10205072 B/op	  150868 allocs/op
PASS
ok  	github.com/cortexproject/cortex/pkg/chunk	7.917s
```

Reply:
So it looks like if there is duplicates the new append with the map unique values  is faster, no surprise here. However if there is no duplicates the old version is faster.

Memory wise I think the increase is negligible. (25B per op).

I will submit the benchmark now.

Reply:
After some discussions with Goutham, let's not take any risk and leave the `parseIndexEntries` as is. Other functions (Fetching labels) are not used by Cortex and should be fine to change.

Reply:
done ! I think we should be good to go here.